File: sdk/golang/internal/infra/default_transport.go
package infra

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"reflect"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/internal/interfaces"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/common/logger"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/types"
	"github.com/google/go-querystring/query"
)

type DefaultTransport struct {
	clientOption    *types.ClientOption
	transportOption *types.TransportOption
	singer          *KcSigner
	httpClient      *http.Client
}

func NewDefaultTransport(clientOption *types.ClientOption) *DefaultTransport {
	option := clientOption.TransportOption
	if option == nil {
		option = types.NewTransportOption()
	}

	if clientOption.Key == "" || clientOption.Secret == "" || clientOption.Passphrase == "" {
		logger.GetLogger().Warnf("no secret information provided, only the public channel API is accessible")
	}

	t := &DefaultTransport{
		clientOption:    clientOption,
		transportOption: option,
		singer: NewKcSigner(clientOption.Key, clientOption.Secret, clientOption.Passphrase,
			clientOption.BrokerName, clientOption.BrokerPartner, clientOption.BrokerKey),
	}
	t.httpClient = &http.Client{
		Transport: t.applyOption(&http.Transport{}),
		Timeout:   t.transportOption.Timeout,
	}
	return t
}

func (t *DefaultTransport) applyOption(transport *http.Transport) *http.Transport {
	transport.DisableKeepAlives = !t.transportOption.KeepAlive
	transport.MaxIdleConns = t.transportOption.MaxIdleConns
	transport.MaxIdleConnsPerHost = t.transportOption.MaxIdleConnsPerHost
	transport.MaxConnsPerHost = t.transportOption.MaxConnsPerHost
	transport.IdleConnTimeout = t.transportOption.IdleConnTimeout
	transport.TLSHandshakeTimeout = t.transportOption.TLSHandshakeTimeout
	transport.DialContext = (&net.Dialer{Timeout: t.transportOption.Timeout}).DialContext
	if t.transportOption.Proxy != nil {
		transport.Proxy = t.transportOption.Proxy
	}
	return transport
}

func (t *DefaultTransport) DoWithRetry(req *http.Request) (resp *http.Response, err error) {
	if t.transportOption.Interceptors != nil {

		for _, interceptor := range t.transportOption.Interceptors {
			req, err = interceptor.Before(req)
			if err != nil {
				return
			}
		}

		defer func() {
			for _, interceptor := range t.transportOption.Interceptors {
				resp, err = interceptor.After(req, resp, err)
				if err != nil {
					return
				}
			}
		}()
	}

	for attempt := 0; attempt <= t.transportOption.MaxRetries; attempt++ {
		resp, err = t.httpClient.Do(req)
		if err == nil {
			return resp, nil
		}

		if attempt < t.transportOption.MaxRetries && t.shouldRetry(err) {
			logger.GetLogger().Errorf("http request failed, reason %s, attemp %d, max %d, retry in %f seconds...", err.Error(), attempt+1, t.transportOption.MaxRetries, t.transportOption.RetryDelay.Seconds())
			time.Sleep(t.transportOption.RetryDelay)
		}
	}

	return nil, errors.New("http request failed after maximum retries, last error: " + err.Error())
}

func (t *DefaultTransport) shouldRetry(err error) bool {
	if err != nil {
		var netErr net.Error
		if errors.As(err, &netErr) && (netErr.Temporary() || netErr.Timeout()) {
			return true
		}
	}

	return false
}

func (t *DefaultTransport) processPathVariable(path string, request interface{}) string {
	hasQueryVariable := false
	variableName := ""
	idx := strings.Index(path, "{")
	if idx != -1 && idx != len(path)-1 {
		path := path[idx+1:]
		idx = strings.Index(path, "}")
		if idx != -1 {
			hasQueryVariable = true
			variableName = path[:idx]
		}
	}

	if hasQueryVariable && request != nil && variableName != "" {
		val := reflect.ValueOf(request)
		if val.Kind() == reflect.Ptr {
			val = val.Elem()
		}
		if val.Kind() == reflect.Struct {
			for i := 0; i < val.Type().NumField(); i++ {
				tag := val.Type().Field(i).Tag.Get("path")
				if tag == variableName {
					fieldValue := val.Field(i)
					if fieldValue.Kind() == reflect.Ptr && !fieldValue.IsNil() {
						fieldValue = fieldValue.Elem()
					}
					variableValue := fmt.Sprintf("%v", fieldValue.Interface())
					path = strings.Replace(path, "{"+variableName+"}", variableValue, -1)
					break
				}
			}
		}
	}

	return path
}

func (t *DefaultTransport) processHeaders(method string, broker bool, rawPath string, body []byte, r *http.Request) {
	r.Header.Set("Content-Type", "application/json")
	r.Header.Set("User-Agent", "Kucoin-Universal-Go-SDK/1.0")
	var b bytes.Buffer
	b.WriteString(method)
	b.WriteString(rawPath)
	b.Write(body)

	if broker {
		h := t.singer.BrokerHeaders(b.String())
		for k, v := range h {
			r.Header.Set(k, v)
		}
	} else {
		h := t.singer.Headers(b.String())
		for k, v := range h {
			r.Header.Set(k, v)
		}
	}
}

func encodeRaw(v url.Values) string {
	if len(v) == 0 {
		return ""
	}
	var buf strings.Builder
	keys := make([]string, 0, len(v))
	for k := range v {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		vs := v[k]
		for _, v := range vs {
			if buf.Len() > 0 {
				buf.WriteByte('&')
			}
			buf.WriteString(k)
			buf.WriteByte('=')
			buf.WriteString(v)
		}
	}
	return buf.String()
}

func (t *DefaultTransport) processRequest(ctx context.Context, broker bool, endpoint, method string,
	path string, request interface{}, requestJson bool) (*http.Request, error) {
	method = strings.ToUpper(method)
	path = t.processPathVariable(path, request)
	rawPath := path

	var reqBody []byte
	var err error

	if requestJson {
		if request != nil {
			reqBody, err = json.Marshal(request)
			if err != nil {
				return nil, err
			}
		}
	} else {
		switch method {
		case http.MethodGet:
			fallthrough
		case http.MethodDelete:
			{
				if request != nil {
					values, err := query.Values(request)
					if err != nil {
						return nil, err
					}
					if len(values) > 0 {
						path += "?" + values.Encode()
						rawPath += "?" + encodeRaw(values)
					}
				}

				break
			}
		case http.MethodPost:
			{
				if request != nil {
					reqBody, err = json.Marshal(request)
					if err != nil {
						return nil, err
					}
				}
				break
			}
		default:
			return nil, fmt.Errorf("invalid method: %s", method)
		}
	}

	fullPath := fmt.Sprintf("%s%s", endpoint, path)
	req, err := http.NewRequestWithContext(ctx, method, fullPath, bytes.NewBuffer(reqBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	t.processHeaders(method, broker, rawPath, reqBody, req)
	return req, nil
}

func (t *DefaultTransport) processLimit(resp *http.Response) *types.RestRateLimit {
	limit, err := strconv.ParseInt(resp.Header.Get("gw-ratelimit-limit"), 10, 64)
	if err != nil {
		limit = -1
	}
	remaining, err := strconv.ParseInt(resp.Header.Get("gw-ratelimit-remaining"), 10, 64)
	if err != nil {
		remaining = -1
	}
	reset, err := strconv.ParseInt(resp.Header.Get("gw-ratelimit-reset"), 10, 64)
	if err != nil {
		reset = -1
	}

	return &types.RestRateLimit{
		Limit:     limit,
		Remaining: remaining,
		Reset:     reset,
	}
}

func (t *DefaultTransport) processResponse(resp *http.Response, responseObj interfaces.Response) error {
	defer resp.Body.Close()
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("failed to read response body: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("invalid status code: %d, msg: %s", resp.StatusCode, respBody)
	}

	commonResponse := &types.RestResponse{}

	err = json.Unmarshal(respBody, commonResponse)
	if err != nil {
		return fmt.Errorf("failed to unmarshal common response: %w", err)
	}

	// fast gc
	respBody = nil

	commonResponse.RateLimit = t.processLimit(resp)
	responseObj.SetCommonResponse(commonResponse)
	err = commonResponse.Error()
	if err != nil {
		return err
	}

	if len(commonResponse.Data) == 0 {
		return nil
	}

	err = json.Unmarshal(commonResponse.Data, responseObj)
	if err != nil {
		return fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return nil
}

func (t *DefaultTransport) Call(ctx context.Context, domain string, broker bool, method string, path string,
	requestObj interface{}, responseObj interfaces.Response, requestJson bool) error {

	endpoint := t.clientOption.SpotEndpoint

	domain = strings.ToLower(domain)

	switch domain {
	case types.DomainTypeSpot:
		{
			endpoint = t.clientOption.SpotEndpoint
			break
		}

	case types.DomainTypeFutures:
		{
			endpoint = t.clientOption.FuturesEndpoint
			break
		}
	case types.DomainTypeBroker:
		{
			endpoint = t.clientOption.BrokerEndpoint
			break
		}
	default:
		return fmt.Errorf("invalid domain: %s", domain)
	}

	req, err := t.processRequest(ctx, broker, endpoint, method, path, requestObj, requestJson)
	if err != nil {
		return err
	}

	resp, err := t.DoWithRetry(req)
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}

	return t.processResponse(resp, responseObj)
}

func (t *DefaultTransport) Close() error {
	t.httpClient.CloseIdleConnections()
	return nil
}

File: sdk/python/kucoin_universal_sdk/internal/infra/default_transport.py
import json
from enum import Enum
from io import BytesIO
from typing import Any
from urllib.parse import urlencode

import requests
from pydantic import BaseModel
from requests.adapters import HTTPAdapter

from kucoin_universal_sdk.model.client_option import ClientOption
from kucoin_universal_sdk.model.common import RestRateLimit, RestError
from kucoin_universal_sdk.model.common import RestResponse
from kucoin_universal_sdk.model.constants import DomainType
from kucoin_universal_sdk.model.transport_option import TransportOption
from .default_signer import KcSigner
from ..interfaces.response import Response
from ..interfaces.transport import Transport

class DefaultTransport(Transport):
    def __init__(self, client_option: ClientOption, sdk_version: str):
        transport_option = client_option.transport_option
        if transport_option is None:
            transport_option = TransportOption()
            client_option.transport_option = transport_option
        self.client_option = client_option
        self.sdk_version = sdk_version

        self.transport_option = transport_option
        self.signer = KcSigner(client_option.key, client_option.secret, client_option.passphrase,
                               client_option.broker_name, client_option.broker_partner, client_option.broker_key)

        self.http_client = self.create_http_client(transport_option)

    def create_http_client(self, option: TransportOption) -> requests.Session:
        session = requests.Session()
        session = self.apply_option(session, option)
        return session

    def apply_option(self, session: requests.Session, option: TransportOption) -> requests.Session:
        adapter = HTTPAdapter(
            max_retries=option.max_retries,
            pool_connections=option.max_connection_per_pool,
            pool_maxsize=option.max_pool_size,
            pool_block=True
        )
        session.mount('https://', adapter)
        session.mount('http://', adapter)

        session.headers.update({
            'Connection': 'keep-alive' if option.keep_alive else 'close'
        })

        if option.proxy:
            session.proxies = option.proxy
        return session

    def process_headers(self, body: bytes, raw_url: str, request: requests.PreparedRequest, method: str, broker: bool):
        request.headers["Content-Type"] = "application/json"
        request.headers["User-Agent"] = f"Kucoin-Universal-Python-SDK/{self.sdk_version}"

        b = BytesIO()
        b.write(method.encode())
        b.write(raw_url.encode())
        b.write(body)
        payload = b.getvalue().decode()

        if broker:
            headers = self.signer.broker_headers(payload)
        else:
            headers = self.signer.headers(payload)

        for key, value in headers.items():
            request.headers[key] = value

    def process_path_variable(self, path: str, request_obj: Any) -> str:
        if request_obj is None:
            return path

        path_variables = {
            field_info.alias or field_name: getattr(request_obj, field_name)
            for field_name, field_info in request_obj.model_fields.items()
            if field_info.json_schema_extra and field_info.json_schema_extra.get("path_variable", False)
        }

        missing_placeholders = [key for key, value in path_variables.items() if value is None]
        if missing_placeholders:
            raise ValueError(f"Missing path variable value(s) for {', '.join(missing_placeholders)}")

        return path.format(**path_variables)

    def raw_query(self, query_dict):
        if not query_dict:
            return ""

        result = []
        for key in query_dict.keys():
            values = query_dict[key]
            if not isinstance(values, list):
                values = [values]
            for value in values:
                result.append(f"{key}={value}")

        return "&".join(result)

    def process_request(self, request_obj: Any, broker: bool, path: str, endpoint: str, method: str,
                        request_as_json: bool, **kwargs: Any) -> requests.Request:
        path = self.process_path_variable(path, request_obj)
        raw_url = path
        req_body = None
        full_path = f"{endpoint}{path}"

        if request_as_json:
            if request_obj:
                req_body = request_obj.to_json()
        else:
            if method in ["GET", "DELETE"]:
                if request_obj:
                    query_dict = {
                        key: value.value if isinstance(value, Enum) else value
                        for key, value in request_obj.to_dict().items()
                    }
                    query_params = urlencode(query_dict)
                    raw_query_params = self.raw_query(query_dict)
                    if query_params:
                        full_path += f"?{query_params}"
                        raw_url += f"?{raw_query_params}"
            elif method == "POST":
                if request_obj:
                    req_body = request_obj.to_json()
            else:
                raise ValueError(f"Invalid method: {method}")

        req = requests.Request(
            method=method,
            url=full_path,
            data=req_body,
        )
        prepared_req = req.prepare()
        self.process_headers(req_body.encode() if req_body is not None else b"", raw_url, prepared_req, method, broker)

        return prepared_req

    def do_with_retry(self, req: requests.Request) -> requests.Response:
        resp = None
        err = None
        try:
            for interceptor in self.transport_option.interceptors:
                try:
                    req = interceptor.before(req)
                except Exception as e:
                    raise Exception(f"Interceptor before failed: {e}") from e

            try:
                resp = self.http_client.send(req, timeout=(
                    self.transport_option.connect_timeout, self.transport_option.read_timeout))
                err = None
            except Exception as e:
                err = e
                raise e
        finally:
            for interceptor in self.transport_option.interceptors:
                try:
                    resp = interceptor.after(req, resp, err)
                except Exception as e:
                    raise Exception(f"Interceptor after failed: {e}") from e

        return resp

    def process_limit(self, resp: requests.Response) -> RestRateLimit:
        limit = int(resp.headers.get("gw-ratelimit-limit", -1))
        remaining = int(resp.headers.get("gw-ratelimit-remaining", -1))
        reset = int(resp.headers.get("gw-ratelimit-reset", -1))

        return RestRateLimit(
            limit=limit,
            remaining=remaining,
            reset=reset
        )

    def process_response(self, resp: requests.Response, response_obj: Response) -> Any:
        resp_body = resp.content
        if resp.status_code != requests.codes.ok:
            raise RestError(None, f"Invalid status code: {resp.status_code}, msg: {resp_body.decode('utf-8')}")

        try:
            common_response = RestResponse.from_json(resp_body.decode('utf-8'))
        except json.JSONDecodeError as e:
            raise RestError(None, f"Failed to parse common response JSON: {str(e)}")

        resp_body = None
        common_response.rate_limit = self.process_limit(resp)

        if common_response.error():
            raise RestError(common_response, "")

        if common_response.data is None:
            response_obj.set_common_response(common_response)
            return response_obj

        try:
            response_obj = response_obj.from_dict(common_response.data)
        except json.JSONDecodeError as e:
            raise RestError(common_response, f"Failed to parse response JSON data: {str(e)}")
        finally:
            response_obj.set_common_response(common_response)
            return response_obj

    def call(self, domain: str, broker: bool, method: str, path: str, request_obj: Any, response_obj: Response,
             request_as_json: bool, **kwargs: Any) -> Any:
        domain = domain.lower()
        method = method.upper()
        if domain == DomainType.SPOT.value:
            endpoint = self.client_option.spot_endpoint
        elif domain == DomainType.FUTURES.value:
            endpoint = self.client_option.futures_endpoint
        elif domain == DomainType.BROKER.value:
            endpoint = self.client_option.broker_endpoint
        else:
            raise ValueError(f"Invalid domain: {domain}")

        if request_obj is not None and not isinstance(request_obj, BaseModel):
            raise TypeError(f"The request object must inherit from BaseModel, got {type(request_obj)}.")

        try:
            path = self.process_path_variable(path, request_obj)
            req = self.process_request(request_obj, path=path, endpoint=endpoint, method=method, broker=broker,
                                       request_as_json=request_as_json, **kwargs)
            response_body = self.do_with_retry(req)
        except Exception as err:
            raise RestError(None, f"Request failed: {err}")
        return self.process_response(response_body, response_obj)

    def close(self):
        self.http_client.close()